pymess/functions.py

```diff
@@ -0,0 +1,55 @@
+from django.db import transaction
+
+
+def create_message(backend_fnc, recipient, content, related_objects, tag, template, **kwargs):
+    """
+    Create message which will be logged in the database.
+    :param backend_fnc: get backend function which will be evaluated based on recipient
+    :param recipient: email or phone number of the recipient
+    :param content: content of the message
+    :param related_objects: list of related objects that will be linked with the message using generic
+    relation
+    :param tag: string mark that will be saved with the message
+    :param template: template object from which content of the message was created
+    :param kwargs: extra attributes that will be saved with the message
+    """
+    return backend_fnc(recipient=recipient).create_message(recipient, content, related_objects, tag, template, **kwargs)
+
+
+def publish_messages(backend_fnc, messages):
+    """
+    Sends more messages together. If concrete backend provides send more messages at once the method
+    can be overridden
+    :param backend_fnc: get backend function which will be evaluated based on recipient
+    :param messages: list of messages
+    """
+    backends_messages_map = {}
+    for message in messages:
+        backend = backend_fnc(recipient=message.recipient)
+        try:
+            backends_messages_map[backend].append(message)
+        except KeyError:
+            backends_messages_map[backend] = [message]
+    for backend, messages in backends_messages_map:
+        backend._publish_messages(messages)
```

Comment:
you shoud not call protected method. publish_messages must be public

Reply:
I wanted to make it clear we don;r want this method to be called from the outside and use this function instead.

Reply:
It's public now. I just changed the name to `bulk_send_messages` in controller and kept also `publish_messages` on backend (to differentiate these two)